docs: Add IJON Documents (#40)
fuzzuf/fuzzuf: docs/algorithms/ijon/algorithm_en.md

# IJON

## What is IJON?

> IJON is an annotation mechanism that allows PUTs to return new types of feedback to fuzzers, and a fuzzer that supports those feedbacks, proposed by SysSec. Many famous fuzzers are classified as coverage-guided fuzzers, which try to find new behaviors in a program by receiving code coverage as feedback from PUTs. A typical coverage-guided fuzzer has the following weaknesses: 

Reviewer: ```suggestion
[IJON](https://github.com/RUB-SysSec/ijon/)[^ijon] is an annotation mechanism that allows PUTs to return new types of feedback to fuzzers and is a fuzzer that supports those feedback, proposed by [SysSec](https://informatik.rub.de/syssec/). Many famous fuzzers are classified as coverage-guided fuzzers, which try to find new behavior in a program by receiving code coverage as feedback from PUTs. A typical coverage-guided fuzzer has the following weaknesses: 
```

Author: 消えていましたSorry

Author: 「アノテーションとファザーのセット」なので、is A and is Bはおかしそう（it is a car and is a busはだいぶおかしい）

Reviewer: Okay. Then, please just fix the other parts.

Reviewer: ```suggestion
[IJON](https://github.com/RUB-SysSec/ijon/)[^ijon] is an annotation mechanism that allows PUTs to return new types of feedback to fuzzers, and a fuzzer that supports those feedback, proposed by [SysSec](https://informatik.rub.de/syssec/). Many famous fuzzers are classified as coverage-guided fuzzers, which try to find new behavior in a program by receiving code coverage as feedback from PUTs. A typical coverage-guided fuzzer has the following weaknesses: 
```

> - They do not care about the order in which the code coverage was obtained. For example, suppose there is a bug whose triggering condition is "function B must be executed immediately after function A is executed". Because the fuzzer cannot distinguish between the input that causes the bug and the input that causes the execution of function A after the execution of function B, it may try only the latter and overlook the bug. More to the point, many algorithms cannot distinguish between "input that causes both function A and function B to be called" and "a set of input that causes function A to be called and input that causes function B to be called" in the first place, and if one of the two is already tested, the other won't be tested.

Reviewer: ```suggestion
- It does not care about the order it obtained the code coverage. For example, suppose there is a bug whose triggering condition is "function B must execute immediately after function A executes." Because the fuzzer cannot distinguish between the input that causes the bug and the one that causes the execution of function A after function B, it may try only the latter and overlook the bug. More to the point, many algorithms cannot distinguish between "the input that causes both function A and function B to be called" and "two inputs such that one of them executes only function A and the other executes only function B." If one of them has been tested already, the fuzzer will not test the other.
```

Author: the orderにin whichは必須だと思うけどどうかな（the number of timesとかとは話が違う気がする？）
executeはあんまり自動詞で使うイメージないけどどうですか？（runなら分かる）
"the input that causes both"は"an input ~"の方が良さそう
two inputsとしてしまうと、"one of them"のthemがtwo inputsなのか"the input ~ and two inputs"なのか不明っぽくて、setにしたい。あるいはone of themの表現何か別の良いやつないですかね

Reviewer: 全体的に日本語をそのまま直したっぽい受動的な文が多くて読みづらかったので、可能な限り主語を明確にしたいと思っています。

> the orderにin whichは必須だと思う

では"the order in which it obtained the code coverage"が妥当かと思います。

> executeはあんまり自動詞で使うイメージない

'whose triggering condition is "executing function B immediately after the execution of function A."'はどうでしょうか。

> two inputsとしてしまうと...

同じことを思いましたが、"a set of A and B"という表現を聞いたことがない上、それぞれが長いので読みづらかったです。"If the latter two inputs are tested first, ... former one"とかはどうでしょうか。

Reviewer: ```suggestion
- It does not care about the order in which it obtained the code coverage. For example, suppose there is a bug whose triggering condition is "executing function B immediately after the execution of function A." Because the fuzzer cannot distinguish between the input that causes the bug and the one that causes the execution of function A after function B, it may try only the latter and overlook the bug. More to the point, many algorithms cannot distinguish between "the input that causes both function A and function B to be called" and "two inputs such that one of them executes only function A and the other executes only function B." If the latter two inputs are tested first, the fuzzer will not test the former one.
```

Author: あっSuggestionこれかあ
after the execution以外は全部賛同したい

```suggestion
- It does not care about the order in which it obtained the code coverage. For example, suppose there is a bug whose triggering condition is "executing function B immediately after executing function A". Because the fuzzer cannot distinguish between the input that causes the bug and the one that causes the execution of function A after function B, it may try only the latter and overlook the bug. More to the point, many algorithms cannot distinguish between "the input that causes both function A and function B to be called" and "two inputs such that one of them executes only function A and the other executes only function B." If the latter two inputs are tested first, the fuzzer will not test the former one.
```

Reviewer: LGTM

>   - Among the types of code coverage, path coverage can deal with this problem to some extent. However, there is a trade-off in that if a fuzzer over-stores inputs with different execution paths, it is more likely to retain similar inputs that result in the same fuzzing result, and the overall efficiency of a fuzzing campaign will be reduced. It is difficult to automatically adjust this trade-off to a good degree.

Reviewer: ```suggestion
  - Among the types of code coverage, path coverage can deal with this problem to some extent. However, there is a trade-off in that if a fuzzer over-stores inputs with different execution paths, it is more likely to retain similar inputs that result in the same fuzzing result, and it will reduce the overall efficiency of the fuzzing campaign. It is difficult to adjust this trade-off automatically.
```

Author: fuzzing campaignは特定のものではないからaかなという気がする。もっと細かいことをいうと、it will reduceにする場合itがなんなのか若干曖昧すぎる気もするので、「文全体がそうさせる」ということで, which will reduceを使いたいかもしれない

Reviewer: ついでに"result in ... result"が気になっていたので直しました。
```suggestion
  - Among the types of code coverage, path coverage can deal with this problem to some extent. However, there is a trade-off in that if a fuzzer over-stores inputs with different execution paths, it is more likely to retain similar inputs that yield the same fuzzing result, which will eventually reduce the overall efficiency of a fuzzing campaign. It is difficult to adjust this trade-off automatically.
```

> - There can be internal state changes that cannot be noticed by code coverage. For example, as described in IJON's paper, consider the coordinates of the player in a game. The position of the player on the screen is likely to be important in discovering new states of the game. If the player is in the upper-left corner of the screen, he may be closer to the coordinates of the new event than if in the lower right corner of the screen. However, if only code coverage is used as feedback, the same feedback will be returned regardless of which coordinate the player is at.

Reviewer: ```suggestion
- There may be some internal state changes that the code coverage cannot notice. For example, as described in IJON's paper, consider the coordinates of the player in a game. The position of the player on the screen is likely to be important in discovering new states of the game. If the player is in the upper-left corner of the screen, he may be closer to the coordinates of the new event than if in the lower right corner of the screen. However, if the fuzzer just uses the code coverage, both of them will produce the same feedback regardless of which coordinate the player is at.
```

Author: これbyを使ってるからそもそも微妙（withの方が良かった）が、code coverageがnoticeするというよりはcode coverageによってnoticeしたいので、cannot revealとかのほうが良いかも

Reviewer: ```suggestion
- There may be some internal state changes that the code coverage cannot reveal. For example, as described in IJON's paper, consider the coordinates of the player in a game. The position of the player on the screen is likely to be important in discovering new states of the game. If the player is in the upper-left corner of the screen, he may be closer to the coordinates of the new event than if in the lower right corner of the screen. However, if the fuzzer just uses the code coverage, both of them will produce the same feedback regardless of which coordinate the player is at.
```

IJON proposes a simple solution to these problems: human annotation on PUTs. When building a PUT from source code and instrumenting it to obtain coverage, humans can add annotations to the source code to customize the feedback that the PUT gives to a fuzzer. There are various annotations provided by IJON that humans can use to specify what they consider to be important internal states. For example, annotations such as "record the maximum value of a variable in the feedback" or "record the minimum difference between two variables" are possible.

> In practice, because IJON is implemented based on AFL, the feedback returned by PUT is (Hashed) Edge Coverage, which is passed on to the fuzzer via shared memory. Therefore, the annotations that can be written in the source code are specifically implemented as functions and macros that write values to the shared memory. These macros and functions are compiled together when the instrumentation tools instrument the Edge Coverage.

Reviewer: ```suggestion
In practice, because IJON is implemented based on AFL, the feedback returned by PUT is (Hashed) Edge Coverage, which is passed on to the fuzzer via shared memory. Therefore, one must design the annotations in the source code as functions and macros that write some values to the shared memory. These macros and functions are compiled together when the instrumentation tools instrument the Edge Coverage.
```

Author: これは意味が変わっていて、もともとはIJONがそう実装しているという意味です（IJON specifically implementedにするなら分かるが、that can be written ~が長いので、ここは受動態のままでいい気はする。そもそもwereにするべきだったというのはありそう

Reviewer: ```suggestion
In practice, because IJON is implemented based on AFL, the feedback returned by PUT is (Hashed) Edge Coverage, which is passed on to the fuzzer via shared memory. Therefore, IJSON specifically implements the annotations, which the user can write in the source code, as functions and macros that write values to the shared memory. These macros and functions are compiled together when the instrumentation tools instrument the Edge Coverage.
```

Author: ```suggestion
In practice, because IJON is implemented based on AFL, the feedback returned by PUT is (Hashed) Edge Coverage, which is passed on to the fuzzer via shared memory. Therefore, IJSON specifically implemented the annotations, which the user can write in the source code, as functions and macros that write values to the shared memory. These macros and functions are compiled together when the instrumentation tools instrument the Edge Coverage.
```

Author: Fixed in https://github.com/fuzzuf/fuzzuf/pull/40/commits/da3558e7d53cc3c7b426308f793f03a71d5ef172

Thus, because IJON is an AFL-based fuzzer and has an interface for harness description required in practical fuzzing, it has been implemented on fuzzuf to improve the applicability of fuzzuf.

## How to use fuzzuf's IJON CLI

With `fuzzuf` installed, run

```bash
fuzzuf ijon -i path/to/initial/seeds/ path/to/PUT @@
```

> to start IJON. The global options that can be specified are the same as for AFL.

Reviewer: ```suggestion
to start IJON. The global options available are the same as for AFL.
```

For AFL options, see [AFL/algorithm_en.md](/docs/algorithms/afl/algorithm_en.md).

> The local options for IJON that can be used are:

Reviewer: ```suggestion
The local option for IJON is:
```

- `--forksrv 0|1`
  - If 1 is specified, then fork server mode is enabled. It is enabled by default.

## Algorithm Overview

IJON is implemented in a way that retains most of the functions of AFL, and adds additional functions. Roughly speaking, the differences from AFL are as follows:

> - Some cases of havoc mutation have been modified.

Reviewer: ```suggestion
- Some cases of havoc mutation are modified.
```

- IJON has its own seed queue, apart from the AFL seed queue.

>   - For each element of a 64-bit non-negative integer array in shared memory, the IJON seed queue saves the seed that makes a program record the largest value in the element.

Reviewer: ```suggestion
  - For each element of the 64-bit non-negative integer array of the shared memory, the IJON seed queue saves the seed that made PUT record the maximum value in the element.
```

Author: これ共有メモリ上に他のデータ構造もあるので、inのままの方が良いかも。一個しかないのでたしかにtheでも良さそうではあるんだけど、複数回fuzzer起動してそれらが別のものを使うイメージを持つと別にaのままでいいかなという気がする。

Reviewer: ```suggestion
  - For each element of a 64-bit non-negative integer array in shared memory, the IJON seed queue saves the seed that made a program record the maximum value in the element.
```

- At the beginning of the fuzzing loop, the procedure branches randomly.
  - 80% of the time, a seed is selected from the IJON seed queue. In this case, the fuzzer immediately moves to the havoc stage, and returns to the beginning of the fuzzing loop after a certain number of havoc mutations.
  - 20% of the time, a seed is selected from the AFL seed queue. In this case, mutation is performed in the same flow as the original AFL.

> - After a PUT is executed, the IJON seed queue is updated based on the feedback obtained from the PUT.

Reviewer: ```suggestion
- After a PUT exits, the IJON seed queue is updated based on the feedback obtained from the PUT.
```

  - Even when AFL is selected in 20% probability, the IJON seed queue is also updated.

> - Some of the constant values have been changed.

Reviewer: ```suggestion
- Some of the constants are changed.
```

 #define _IJON_CONCAT(x, y) x##y
-#define _IJON_UNIQ_NAME() IJON_CONCAT(temp,__LINE__)
+#define _IJON_UNIQ_NAME IJON_CONCAT(temp,__LINE__)
 #define _IJON_ABS_DIST(x,y) ((x)<(y) ? (y)-(x) : (x)-(y))

 #define IJON_BITS(x) ((x==0)?{0}:__builtin_clz(x))
 #define IJON_INC(x) ijon_map_inc(ijon_hashstr(__LINE__,__FILE__)^(x))
 #define IJON_SET(x) ijon_map_set(ijon_hashstr(__LINE__,__FILE__)^(x))

-#define IJON_CTX(x) ({ uint32_t hash = hashstr(__LINE__,__FILE__); ijon_xor_state(hash); __typeof__(x) IJON_UNIQ_NAME() = (x); ijon_xor_state(hash); IJON_UNIQ_NAME(); })
+#define IJON_CTX(x) ({ uint32_t hash = ijon_hashstr(__LINE__,__FILE__); ijon_xor_state(hash); __typeof__(x) IJON_UNIQ_NAME = (x); ijon_xor_state(hash); IJON_UNIQ_NAME; })

 #define IJON_MAX(x) ijon_max(ijon_hashstr(__LINE__,__FILE__),(x))
 #define IJON_MIN(x) ijon_max(ijon_hashstr(__LINE__,__FILE__),0xffffffffffffffff-(x))
diff --git a/test.c b/test.c
index 50b1b05..aa022f6 100644
--- a/test.c
+++ b/test.c
@@ -3,6 +3,7 @@
 #include<assert.h>
 #include<stdbool.h>
 #include <stdlib.h>
+#include <stdint.h>
```

For example, you can build test.c and fuzz the produced binary with the following commands:

```bash

> $ (path_to_ijon)/llvm_mode/afl-clang-fast (path_to_ijon)/test.c -o test

Reviewer: As far as I checked, `afl-clang-fast` is created at the root directory of IJON if you build it as instructed in [this section](https://github.com/RUB-SysSec/ijon/#compile-aflijon).
```suggestion
$ (path_to_ijon)/afl-clang-fast (path_to_ijon)/test.c -o test
```

Author: That's definitely right

 #define IJON_INC(x) ijon_map_inc(ijon_hashstr(__LINE__,__FILE__)^(x))
 #define IJON_SET(x) ijon_map_set(ijon_hashstr(__LINE__,__FILE__)^(x))

-#define IJON_CTX(x) ({ uint32_t hash = hashstr(__LINE__,__FILE__); ijon_xor_state(hash); __typeof__(x) IJON_UNIQ_NAME() = (x); ijon_xor_state(hash); IJON_UNIQ_NAME(); })
+#define IJON_CTX(x) ({ uint32_t hash = ijon_hashstr(__LINE__,__FILE__); ijon_xor_state(hash); __typeof__(x) IJON_UNIQ_NAME = (x); ijon_xor_state(hash); IJON_UNIQ_NAME; })

 #define IJON_MAX(x) ijon_max(ijon_hashstr(__LINE__,__FILE__),(x))
 #define IJON_MIN(x) ijon_max(ijon_hashstr(__LINE__,__FILE__),0xffffffffffffffff-(x))
diff --git a/test.c b/test.c
index 50b1b05..aa022f6 100644
--- a/test.c
+++ b/test.c
@@ -3,6 +3,7 @@
 #include<assert.h>
 #include<stdbool.h>
 #include <stdlib.h>
+#include <stdint.h>

 #define compare(x,y) IJON_CTX(compare_w((x),(y)))
 bool compare_w(int x, int y){
```

For example, you can build test.c and fuzz the produced binary with the following commands:

```bash
$ (path_to_ijon)/llvm_mode/afl-clang-fast (path_to_ijon)/test.c -o test
$ mkdir /tmp/ijon_test_indir/ && echo hello > /tmp/ijon_test_indir/hello

> $ fuzzuf ijon -i /tmp/ijon_test_indir/ ./test

Reviewer: Please describe what is supposed to happen if I run this command.
e.g., "It is successful if fuzzuf finds a bug in about X minutes."
FYI I could find a unique crash in about 3 minutes in my environment.

It's also better to describe the difference between AFL. AFL also found the bug about 30 seconds later IJON found it in my env. If IJON is likely to find the bug faster than AFL, please explain it in the document.

Author: Added in https://github.com/fuzzuf/fuzzuf/pull/40/commits/8aa518ce8013750aa68a9a7b6ca383a7384b2787 .
Here, I don't want to compare it with AFL because the difference between them doesn't look significant. In a single run, AFL may find a crash earlier than IJON(it actually did in my test). Users have to run them many times to be convinced that IJON is faster, which is nevertheless time-consuming. Maybe the difference can be vivid in test2.c, but IJON couldn't find a crash of it within 10 minutes as far as I tested and thus it may be also time-consuming.

Author: In that sense, I think test.c and test2.c are not good exemplifications of IJON's superiority to AFL, while they are nice examples of how to use it. The superiority should be clear when we test them with Super Mario Bros.
